Restore failure due to `The process cannot access the file x because it is being used by another process`

### Description

When building runtime with mono-flavored runtime, restore often fails in CI environment with `The process cannot access the file x because it is being used by another process`. This affects runtime, as well as other builds (confirmed roslyn), thus specificity to runtime is that this does not occur on coreclr-flavored runtime.

### Reproduction Steps
In an Alpine Edge `linux-musl-x64` environment, you can use [this modified aport](https://gitlab.alpinelinux.org/ayakael/aports/-/tree/dotnet7/mono-restore/testing/dotnet7-stage0) to reproduce bug. Following steps to reproduce:
```bash
git clone https://gitlab.alpinelinux.org/ayakael/aports -b dotnet7/mono-restore
cd  aports/testing/dotnet7-stage0
abuild deps unpack prepare build
```
It should eventually fail.

The aport builds a minimum set of components (runtime-mono, roslyn, sdk, aspnetcore, installer) to be able to build an SDK tar, then using that produced tarball with mono-flavored runtime it builds the whole stack again. This aport is usually used to crossbuild to other platforms, but in this case I am using it to easily reproduce the bug.

You'd likely be able to reproduce this on `linux-x64 by building runtime with `/p:PrimaryRuntimeFlavor=Mono` and trying to build runtime with produced artifacts.

### Expected behavior

Restore should occur without issue

### Actual behavior

Restore fails with the following error:
```
/builds/ayakael/aports/testing/dotnet7-stage0/src/bootstrap/sdk/7.0.100-rtm.22519.39/NuGet.RestoreEx.targets(19,5): error : The process cannot access the file '/builds/ayakael/aports/testing/dotnet7-stage0/src/dotnet-d41bfecf5090e9163aa2da251246abed9e756e53/src/runtime/artifacts/obj/System.IO.Ports' because it is being used by another process. [/builds/ayakael/aports/testing/dotnet7-stage0/src/dotnet-d41bfecf5090e9163aa2da251246abed9e756e53/src/runtime/Build.proj]
/builds/ayakael/aports/testing/dotnet7-stage0/src/bootstrap/sdk/7.0.100-rtm.22519.39/NuGet.RestoreEx.targets(19,5): error :    at System.IO.FileSystem.CreateDirectory(String fullPath, UnixFileMode unixCreateMode) [/builds/ayakael/aports/testing/dotnet7-stage0/src/dotnet-d41bfecf5090e9163aa2da251246abed9e756e53/src/runtime/Build.proj]
/builds/ayakael/aports/testing/dotnet7-stage0/src/bootstrap/sdk/7.0.100-rtm.22519.39/NuGet.RestoreEx.targets(19,5): error :    at System.IO.FileSystem.CreateDirectory(String fullPath) [/builds/ayakael/aports/testing/dotnet7-stage0/src/dotnet-d41bfecf5090e9163aa2da251246abed9e756e53/src/runtime/Build.proj]
/builds/ayakael/aports/testing/dotnet7-stage0/src/bootstrap/sdk/7.0.100-rtm.22519.39/NuGet.RestoreEx.targets(19,5): error :    at System.IO.Directory.CreateDirectory(String path) [/builds/ayakael/aports/testing/dotnet7-stage0/src/dotnet-d41bfecf5090e9163aa2da251246abed9e756e53/src/runtime/Build.proj]
/builds/ayakael/aports/testing/dotnet7-stage0/src/bootstrap/sdk/7.0.100-rtm.22519.39/NuGet.RestoreEx.targets(19,5): error :    at NuGet.Commands.BuildAssetsUtils.WriteFiles(IEnumerable`1 files, ILogger log) [/builds/ayakael/aports/testing/dotnet7-stage0/src/dotnet-d41bfecf5090e9163aa2da251246abed9e756e53/src/runtime/Build.proj]
/builds/ayakael/aports/testing/dotnet7-stage0/src/bootstrap/sdk/7.0.100-rtm.22519.39/NuGet.RestoreEx.targets(19,5): error :    at NuGet.Commands.RestoreResult.CommitAssetsFileAsync(LockFileFormat lockFileFormat, IRestoreResult result, ILogger log, Boolean toolCommit, CancellationToken token) [/builds/ayakael/aports/testing/dotnet7-stage0/src/dotnet-d41bfecf5090e9163aa2da251246abed9e756e53/src/runtime/Build.proj]
/builds/ayakael/aports/testing/dotnet7-stage0/src/bootstrap/sdk/7.0.100-rtm.22519.39/NuGet.RestoreEx.targets(19,5): error :    at NuGet.Commands.RestoreResult.CommitAsync(ILogger log, CancellationToken token) [/builds/ayakael/aports/testing/dotnet7-stage0/src/dotnet-d41bfecf5090e9163aa2da251246abed9e756e53/src/runtime/Build.proj]
/builds/ayakael/aports/testing/dotnet7-stage0/src/bootstrap/sdk/7.0.100-rtm.22519.39/NuGet.RestoreEx.targets(19,5): error :    at NuGet.Commands.RestoreRunner.CommitAsync(RestoreResultPair restoreResult, CancellationToken token) [/builds/ayakael/aports/testing/dotnet7-stage0/src/dotnet-d41bfecf5090e9163aa2da251246abed9e756e53/src/runtime/Build.proj]
/builds/ayakael/aports/testing/dotnet7-stage0/src/bootstrap/sdk/7.0.100-rtm.22519.39/NuGet.RestoreEx.targets(19,5): error :    at NuGet.Commands.RestoreRunner.ExecuteAndCommitAsync(RestoreSummaryRequest summaryRequest, IRestoreProgressReporter progressReporter, CancellationToken token) [/builds/ayakael/aports/testing/dotnet7-stage0/src/dotnet-d41bfecf5090e9163aa2da251246abed9e756e53/src/runtime/Build.proj]
/builds/ayakael/aports/testing/dotnet7-stage0/src/bootstrap/sdk/7.0.100-rtm.22519.39/NuGet.RestoreEx.targets(19,5): error :    at NuGet.Commands.RestoreRunner.CompleteTaskAsync(List`1 restoreTasks) [/builds/ayakael/aports/testing/dotnet7-stage0/src/dotnet-d41bfecf5090e9163aa2da251246abed9e756e53/src/runtime/Build.proj]
/builds/ayakael/aports/testing/dotnet7-stage0/src/bootstrap/sdk/7.0.100-rtm.22519.39/NuGet.RestoreEx.targets(19,5): error :    at NuGet.Commands.RestoreRunner.RunAsync(IEnumerable`1 restoreRequests, RestoreArgs restoreArgs, CancellationToken token) [/builds/ayakael/aports/testing/dotnet7-stage0/src/dotnet-d41bfecf5090e9163aa2da251246abed9e756e53/src/runtime/Build.proj]
/builds/ayakael/aports/testing/dotnet7-stage0/src/bootstrap/sdk/7.0.100-rtm.22519.39/NuGet.RestoreEx.targets(19,5): error :    at NuGet.Commands.RestoreRunner.RunAsync(RestoreArgs restoreContext, CancellationToken token) [/builds/ayakael/aports/testing/dotnet7-stage0/src/dotnet-d41bfecf5090e9163aa2da251246abed9e756e53/src/runtime/Build.proj]
/builds/ayakael/aports/testing/dotnet7-stage0/src/bootstrap/sdk/7.0.100-rtm.22519.39/NuGet.RestoreEx.targets(19,5): error :    at NuGet.Build.Tasks.BuildTasksUtility.RestoreAsync(DependencyGraphSpec dependencyGraphSpec, Boolean interactive, Boolean recursive, Boolean noCache, Boolean ignoreFailedSources, Boolean disableParallel, Boolean force, Boolean forceEvaluate, Boolean hideWarningsAndErrors, Boolean restorePC, Boolean cleanupAssetsForUnsupportedProjects, ILogger log, CancellationToken cancellationToken) [/builds/ayakael/aports/testing/dotnet7-stage0/src/dotnet-d41bfecf5090e9163aa2da251246abed9e756e53/src/runtime/Build.proj]
/builds/ayakael/aports/testing/dotnet7-stage0/src/bootstrap/sdk/7.0.100-rtm.22519.39/NuGet.RestoreEx.targets(19,5): error :    at NuGet.Build.Tasks.Console.MSBuildStaticGraphRestore.RestoreAsync(String entryProjectFilePath, IDictionary`2 globalProperties, IReadOnlyDictionary`2 options) [/builds/ayakael/aports/testing/dotnet7-stage0/src/dotnet-d41bfecf5090e9163aa2da251246abed9e756e53/src/runtime/Build.proj]
```

### Regression?

This error is confirmed not present with .NET runtime 6.0.10 in `s390x`, thus it seems to be a regression.

### Known Workarounds

None so far.

### Configuration

Version: SDK `7.0.100-rtm.22519.39`
OS: Alpine LInux Edge docker-based CI Pipelines (although occurs to a lesser degree in s390x VM)
Architecure: Confirmed reproducible in s390x, ppc64le and x64 (with mono-flavored runtime)
Specificity: specific to mono on all platforms


### Other information

logs
ppc64le with runtime: [link](https://gitlab.alpinelinux.org/ayakael/aports/-/jobs/882490/raw)
s390x with runtime: [link](https://gitlab.alpinelinux.org/ayakael/aports/-/jobs/880188/raw)
s390x with roslyn: [link](https://gitlab.alpinelinux.org/ayakael/aports/-/jobs/880132/raw)

Never mind, indeed it happens less, but I still occurs enough to make our build fails consistently:
```
/builds/ayakael/aports/testing/dotnet7-build/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/Tools/source-built/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildArcadeBuild.targets(116,5): error MSB3026: Could not copy "/builds/ayakael/aports/testing/dotnet7-build/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/fsharp/vsintegration/src/FSharp.ProjectSystem.PropertyPages/PropertyPages/xlf/ApplicationPropPage.de.xlf" to "/builds/ayakael/aports/testing/dotnet7-build/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/fsharp/artifacts/source-build/self/src/vsintegration/src/FSharp.ProjectSystem.PropertyPages/PropertyPages/xlf/ApplicationPropPage.de.xlf". Beginning retry 1 in 1000ms. The process cannot access the file '/builds/ayakael/aports/testing/dotnet7-build/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/fsharp/artifacts/source-build/self/src/vsintegration/src/FSharp.ProjectSystem.PropertyPages/PropertyPages/xlf' because it is being used by another process.  [/builds/ayakael/aports/testing/dotnet7-build/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/Tools/source-built/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
```

--- a/src/mono/mono/utils/mono-threads-coop.c
+++ b/src/mono/mono/utils/mono-threads-coop.c
@@ -337,10 +337,10 @@ mono_threads_exit_gc_safe_region_internal (gpointer cookie, MonoStackData *stack
 		return;
 
 #ifdef ENABLE_CHECKED_BUILD_GC
-	W32_DEFINE_LAST_ERROR_RESTORE_POINT;
+	MONO_DEFINE_LAST_ERROR_RESTORE_POINT;
 	if (mono_check_mode_enabled (MONO_CHECK_MODE_GC))
 		coop_tls_pop (cookie);
-	W32_RESTORE_LAST_ERROR_FROM_RESTORE_POINT;
+	MONO_RESTORE_LAST_ERROR_FROM_RESTORE_POINT;
 #endif
 
 	mono_threads_exit_gc_safe_region_unbalanced_internal (cookie, stackdata);
@@ -365,7 +365,7 @@ mono_threads_exit_gc_safe_region_unbalanced_internal (gpointer cookie, MonoStack
 	/* Common to use enter/exit gc safe around OS API's affecting last error. */
 	/* This method can call OS API's that will reset last error on some platforms. */
 	/* To reduce errors, we need to restore last error before exit gc safe. */
-	W32_DEFINE_LAST_ERROR_RESTORE_POINT;
+	MONO_DEFINE_LAST_ERROR_RESTORE_POINT;
 
 	info = (MonoThreadInfo *)cookie;
 
@@ -398,7 +398,7 @@ mono_threads_exit_gc_safe_region_unbalanced_internal (gpointer cookie, MonoStack
 		info->user_data = NULL;
 	}
 
-	W32_RESTORE_LAST_ERROR_FROM_RESTORE_POINT;
+	MONO_RESTORE_LAST_ERROR_FROM_RESTORE_POINT;
 }
 
 void
@@ -652,14 +652,14 @@ mono_threads_suspend_policy_init (void)
 		// otherwise if one of the old environment variables is set, use that.
 		// otherwise use full preemptive suspend.
 
-		W32_DEFINE_LAST_ERROR_RESTORE_POINT;
+		MONO_DEFINE_LAST_ERROR_RESTORE_POINT;
 
 		   (policy = threads_suspend_policy_getenv ())
 		|| (policy = threads_suspend_policy_default ())
 		|| (policy = threads_suspend_policy_getenv_compat ())
 		|| (policy = MONO_THREADS_SUSPEND_FULL_PREEMPTIVE);
 
-		W32_RESTORE_LAST_ERROR_FROM_RESTORE_POINT;
+		MONO_RESTORE_LAST_ERROR_FROM_RESTORE_POINT;
 
 		g_assert (policy);
 		mono_threads_suspend_policy_hidden_dont_modify = (char)policy;

--- a/src/mono/mono/utils/mono-threads.c
+++ b/src/mono/mono/utils/mono-threads.c
@@ -1938,7 +1938,7 @@ mono_thread_info_uninstall_interrupt (gboolean *interrupted)
 	/* Common to uninstall interrupt handler around OS API's affecting last error. */
 	/* This method could call OS API's on some platforms that will reset last error so make sure to restore */
 	/* last error before exit. */
-	W32_DEFINE_LAST_ERROR_RESTORE_POINT;
+	MONO_DEFINE_LAST_ERROR_RESTORE_POINT;
 
 	g_assert (interrupted);
 	*interrupted = FALSE;
@@ -1961,7 +1961,7 @@ mono_thread_info_uninstall_interrupt (gboolean *interrupted)
 	THREADS_INTERRUPT_DEBUG ("interrupt uninstall  tid %p previous_token %p interrupted %s\n",
 		mono_thread_info_get_tid (info), previous_token, *interrupted ? "TRUE" : "FALSE");
 
-	W32_RESTORE_LAST_ERROR_FROM_RESTORE_POINT;
+	MONO_RESTORE_LAST_ERROR_FROM_RESTORE_POINT;
 }
 
 static MonoThreadInfoInterruptToken*

--- a/src/mono/mono/utils/mono-threads.h
+++ b/src/mono/mono/utils/mono-threads.h
@@ -871,6 +871,17 @@ mono_win32_interrupt_wait (PVOID thread_info, HANDLE native_thread_handle, DWORD
 void
 mono_win32_abort_blocking_io_call (THREAD_INFO_TYPE *info);
 
+#else
+
+
+#endif
+
+#ifdef USE_WINDOWS_BACKEND
+
+/* APC calls can change GetLastError while a thread is suspended.  Save/restore it when doing thread
+   state transitions (for example in m2n wrappers) in order to protect the result of the last
+   pinvoke */
+
 #define W32_DEFINE_LAST_ERROR_RESTORE_POINT \
 	const DWORD _last_error_restore_point = GetLastError ();
 
@@ -879,11 +890,31 @@ mono_win32_abort_blocking_io_call (THREAD_INFO_TYPE *info);
 		if (GetLastError () != _last_error_restore_point) \
 			mono_SetLastError (_last_error_restore_point);
 
-#else
+#elif defined(USE_WASM_BACKEND) || defined (USE_POSIX_BACKEND)
+
+#define W32_DEFINE_LAST_ERROR_RESTORE_POINT     \
+	int _last_errno_restore_point = errno;
+
+#define W32_RESTORE_LAST_ERROR_FROM_RESTORE_POINT       \
+        if (errno != _last_errno_restore_point)         \
+                errno = _last_errno_restore_point;
+
+/* Posix semaphores set errno on failure and sporadic wakeup.  GC state transitions are done in n2m
+ * and m2n wrappers and may change the value of errno from the last pinvoke.  Use these macros to
+ * save/restore errno when doing thread state transitions. */
+
+#elif defined(USE_MACH_BACKEND)
+
+/* Mach semaphores don't set errno on failure.  Change this to be the same as POSIX if some other primitives used
+   in thread state transitions pollute errno. */
 
 #define W32_DEFINE_LAST_ERROR_RESTORE_POINT /* nothing */
 #define W32_RESTORE_LAST_ERROR_FROM_RESTORE_POINT /* nothing */
 
+#else
+#error "unknown threads backend, not sure how to save/restore last error"
 #endif
+
+
 
 #endif /* __MONO_THREADS_H__ */

--- a/src/mono/mono/utils/mono-threads.h
+++ b/src/mono/mono/utils/mono-threads.h
@@ -882,20 +882,20 @@ mono_win32_abort_blocking_io_call (THREAD_INFO_TYPE *info);
    state transitions (for example in m2n wrappers) in order to protect the result of the last
    pinvoke */
 
-#define W32_DEFINE_LAST_ERROR_RESTORE_POINT \
+#define MONO_DEFINE_LAST_ERROR_RESTORE_POINT \
 	const DWORD _last_error_restore_point = GetLastError ();
 
-#define W32_RESTORE_LAST_ERROR_FROM_RESTORE_POINT \
+#define MONO_RESTORE_LAST_ERROR_FROM_RESTORE_POINT \
 		/* Only restore if changed to prevent unnecessary writes. */ \
 		if (GetLastError () != _last_error_restore_point) \
 			mono_SetLastError (_last_error_restore_point);
 
 #elif defined(USE_WASM_BACKEND) || defined (USE_POSIX_BACKEND)
 
-#define W32_DEFINE_LAST_ERROR_RESTORE_POINT     \
+#define MONO_DEFINE_LAST_ERROR_RESTORE_POINT     \
 	int _last_errno_restore_point = errno;
 
-#define W32_RESTORE_LAST_ERROR_FROM_RESTORE_POINT       \
+#define MONO_RESTORE_LAST_ERROR_FROM_RESTORE_POINT       \
         if (errno != _last_errno_restore_point)         \
                 errno = _last_errno_restore_point;
 
@@ -908,8 +908,8 @@ mono_win32_abort_blocking_io_call (THREAD_INFO_TYPE *info);
 /* Mach semaphores don't set errno on failure.  Change this to be the same as POSIX if some other primitives used
    in thread state transitions pollute errno. */
 
-#define W32_DEFINE_LAST_ERROR_RESTORE_POINT /* nothing */
-#define W32_RESTORE_LAST_ERROR_FROM_RESTORE_POINT /* nothing */
+#define MONO_DEFINE_LAST_ERROR_RESTORE_POINT /* nothing */
+#define MONO_RESTORE_LAST_ERROR_FROM_RESTORE_POINT /* nothing */
 
 #else
 #error "unknown threads backend, not sure how to save/restore last error"

